gam/clustering.py

```diff
@@ -739,11 +834,126 @@ def _find_medoids(
 
         return centers[i]
 
+    def _dask_find_medoids(
+        self, X, n_clusters, dist_func, centers, verbose, n_samples, delta, i
+    ):
+        """Finding all of the medoids
+
+        Args:
+            X (da.array): The Dataset.
+            n_clusters (int): The number of clusters.
+            dist_func (callable): The distance function.
+            centers (np.ndarray): The centers of the different clusters
+            verbose (bool): Print out updates
+            n_samples (int): The number of samples in the dataset.
+            delta (float): The threshold determining whether or not a value is going to be a part of a cluster.
+            i (int): The index of the center
+
+        Returns:
+            centers (np.ndarray): The list of centers for the different clusters.
+        """
+        # find all medoids
+        mu_x = np.zeros((n_samples))
+        sigma_x = np.zeros((n_samples))
+        if i == 0:
+            d_nearest = np.ones((n_samples))
+        else:
+            d_nearest = np.partition(self.D, 0)[:, 0]
+
+        # available candidates - S_tar - we draw samples from this population
+        unselected_ids = np.arange(n_samples)
+        unselected_ids = np.delete(unselected_ids, centers[0:i])
+        solution_ids = np.copy(unselected_ids)
+
+        # # available candidates - S_tar - we draw samples from this population
+        solution_ids = np.copy(unselected_ids)
+        n_used_ref = 0
+        while (n_used_ref < n_samples) and (solution_ids.shape[0] > 1):
+            # sample a batch from S_ref (for init, S_ref = X)
+            idx_ref = np.random.choice(
+                unselected_ids, size=self.batchsize, replace=True
+            )
+            ci_scale = math.sqrt(
+                (2 * math.log(1.0 / delta)) / (n_used_ref + self.batchsize)
+            )
+            outputs = da.apply_along_axis(
+                lambda j: self._dask_looping_solution_ids(
+                    np.array(X),
+                    idx_ref,
+                    dist_func,
+                    d_nearest,
+                    n_used_ref,
+                    mu_x,
+                    sigma_x,
+                    j,
+                    i,
+                ),
+                arr=solution_ids.reshape(-1, 1),
+                axis=1,
+            ).compute()
```

Comment:
I'm not sure `da.apply_along_axis` makes sense here because you are still converting `X` to a numpy array, so not saving any memory. Can you just use a simple for loop?

Reply:
I could just use numpy.vectorize like the other implementation that I had

Reply:
yeah I think that might more sense, especially if you can keep `X` as a dask array.